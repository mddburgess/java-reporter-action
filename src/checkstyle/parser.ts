import { XmlEntities } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import CheckstyleReport, { CheckstyleSeverity } from "./types";

export default class CheckstyleParser extends ReportParser<CheckstyleReport> {
  private filePath = "";

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "file":
        this.onFileOpen(parseAttrs(tag.attrs) as FileAttrs);
        break;
      case "error":
        this.onErrorOpen(parseAttrs(tag.attrs) as ErrorAttrs);
        break;
    }
  }

  constructor() {
    super({
      violations: [],
    });
  }

  private onFileOpen(attrs: FileAttrs) {
    this.filePath = XmlEntities.decode(attrs.name);
  }

  private onErrorOpen(attrs: ErrorAttrs) {
    this.report.violations.push({
      filePath: this.filePath,
      line: Number(attrs.line),
      column: Number(attrs.column) || 0,
      rule: XmlEntities.decode(attrs.source),
      severity: attrs.severity,
      message: XmlEntities.decode(attrs.message),
    });
  }

  protected onTagClose(tag: TagCloseNode): void {
    // do nothing
  }

  protected onText(tag: TextNode | CDATANode): void {
    // do nothing
  }
}

interface FileAttrs {
  name: string;
}

interface ErrorAttrs {
  line: string;
  column?: string;
  severity: CheckstyleSeverity;
  message: string;
  source: string;
}
