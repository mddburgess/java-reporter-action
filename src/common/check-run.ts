import * as core from '@actions/core';
import * as github from '@actions/github';
import {Annotation} from './github';

class CheckRun {

    private readonly token;
    private readonly octokit;

    private readonly name;
    private checkRunId?: number;

    constructor(name: string) {
        this.token = core.getInput('github-token', {required: true});
        this.octokit = github.getOctokit(this.token);
        this.name = name;
    }

    public async conclude(request: CheckRunRequest) {
        if (request.output) {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion,
                output: {
                    title: request.output.title,
                    summary: request.output.summary
                }
            });
            await this.annotate(request.output);
        } else {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion
            });
        }
    }

    private async annotate(output: CheckOutput) {
        if (!output.annotations || output.annotations.length === 0) {
            return;
        }
        for (let i = 0; i < output.annotations.length; i += 50) {
            await this.updateCheckRun({
                output: {
                    title: output.title,
                    summary: output.summary,
                    annotations: output.annotations.slice(i, i + 50)
                }
            });
        }
    }

    private async createOrUpdateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            await this.createCheckRun(request);
        } else {
            await this.updateCheckRun(request);
        }
    }

    private async createCheckRun(request: CheckRunRequest) {
        core.debug(`Creating ${this.name} check run...`);
        const response = await this.octokit.checks.create({
            ...github.context.repo,
            name: this.name,
            head_sha: this.resolveHeadSha(),
            ...request
        });
        this.checkRunId = response.data.id;
    }

    private async updateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            throw Error('Cannot update a check run before creating it');
        }

        core.debug(`Updating ${this.name} check run...`);
        await this.octokit.checks.update({
            ...github.context.repo,
            check_run_id: this.checkRunId,
            ...request
        });
    }

    private resolveHeadSha() {
        return github.context.payload.pull_request
            ? github.context.payload.pull_request.head.sha
            : github.context.sha;
    }
}

interface CheckRunRequest {
    status?: CheckStatus,
    conclusion?: CheckConclusion,
    output?: CheckOutput
}

type CheckStatus =
    | 'queued'
    | 'in_progress'
    | 'completed';

type CheckConclusion =
    | 'success'
    | 'failure'
    | 'neutral'
    | 'cancelled'
    | 'skipped'
    | 'timed_out'
    | 'action_required';

interface CheckOutput {
    title: string,
    summary: string,
    text?: string,
    annotations?: Annotation[];
}

export default CheckRun;
