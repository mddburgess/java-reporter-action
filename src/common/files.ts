import * as glob from '@actions/glob';
import {Logger} from 'tslog';

const log = new Logger({
    displayFunctionName: false
});

export async function findFiles(searchPaths: string) {
    const patterns = searchPaths.split(',');
    log.debug('Searching paths', patterns);

    const globber = await glob.create(patterns.join('\n'));
    const paths = await globber.glob();

    log.debug(`Found ${paths.length} paths`, paths);
    return paths;
}

export async function findFile(searchPath: string) {
    const paths = await findFiles(searchPath);
    return paths[0];
}
