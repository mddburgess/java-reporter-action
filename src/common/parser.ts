import * as core from "@actions/core";
import * as fs from "fs";
import { Saxophone } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";

export default abstract class ReportParser<T> {
  private readonly context: string[] = ["root"];
  private readonly parser = new Saxophone()
    .on("tagOpen", (tag) => this.handleTagOpen(tag))
    .on("tagClose", (tag) => this.handleTagClose(tag))
    .on("text", (tag) => this.handleText(tag))
    .on("cdata", (tag) => this.handleText(tag));

  protected constructor(protected readonly report: T, private readonly reportPath: string) {}

  read() {
    try {
      const xml = fs.readFileSync(this.reportPath, { encoding: "utf-8" });
      this.parser.parse(xml);
      return this.report;
    } catch (error) {
      core.warning(`Failed to read report: ${this.reportPath}`);
      return undefined;
    }
  }

  protected getContext() {
    return this.context[0];
  }

  private handleTagOpen(tag: TagOpenNode) {
    this.onTagOpen(tag);
    if (tag.isSelfClosing) {
      this.onTagClose(tag);
    } else {
      this.context.unshift(tag.name);
    }
  }

  private handleTagClose(tag: TagCloseNode) {
    this.context.shift();
    this.onTagClose(tag);
  }

  private handleText(tag: TextNode | CDATANode) {
    this.onText(tag);
  }

  protected abstract onTagOpen(tag: TagOpenNode): void;
  protected abstract onTagClose(tag: TagCloseNode): void;
  protected abstract onText(tag: TextNode | CDATANode): void;
}
