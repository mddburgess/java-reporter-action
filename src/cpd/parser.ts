import { XmlEntities } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import CpdReport, { CpdDuplication } from "./types";

export default class CpdParser extends ReportParser<CpdReport> {
  private duplication: CpdDuplication = {
    lines: 0,
    tokens: 0,
    files: [],
  };

  constructor() {
    super({
      duplications: [],
    });
  }

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "duplication":
        this.onDuplicationOpen(parseAttrs(tag.attrs) as DuplicationAttrs);
        break;
      case "file":
        this.onFileOpen(parseAttrs(tag.attrs) as FileAttrs);
        break;
    }
  }

  private onDuplicationOpen(attrs: DuplicationAttrs) {
    this.duplication = {
      lines: Number(attrs.lines),
      tokens: Number(attrs.tokens),
      files: [],
    };
    this.report && this.report.duplications.push(this.duplication);
  }

  private onFileOpen(attrs: FileAttrs) {
    this.duplication.files.push({
      path: XmlEntities.decode(attrs.path),
      startLine: Number(attrs.line),
      endLine: Number(attrs.endline),
      startColumn: Number(attrs.column),
      endColumn: Number(attrs.endcolumn),
    });
  }

  protected onTagClose(tag: TagCloseNode): void {
    // do nothing
  }

  protected onText(tag: TextNode | CDATANode): void {
    // do nothing
  }
}

interface DuplicationAttrs {
  lines: string;
  tokens: string;
}

interface FileAttrs {
  path: string;
  line: string;
  endline: string;
  column: string;
  endcolumn: string;
}
