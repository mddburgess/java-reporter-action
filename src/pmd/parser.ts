import { decode } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import PmdReport, { PmdViolation } from "./types";

export default class PmdParser extends ReportParser<PmdReport> {
  private filePath = "";
  private violation?: PmdViolation;

  constructor(reportPath: string) {
    super(
      {
        violations: [],
      },
      reportPath
    );
  }

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "file":
        this.onFileOpen(parseAttrs(tag.attrs) as FileAttrs);
        break;
      case "violation":
        this.onViolationOpen(parseAttrs(tag.attrs) as ViolationAttrs);
        break;
    }
  }

  private onFileOpen(attrs: FileAttrs) {
    this.filePath = decode(attrs.name, { level: "xml" });
  }

  private onViolationOpen(attrs: ViolationAttrs) {
    this.violation = {
      filePath: this.filePath,
      startLine: Number(attrs.beginline),
      endLine: Number(attrs.endline),
      startColumn: Number(attrs.begincolumn),
      endColumn: Number(attrs.endcolumn),
      ruleset: decode(attrs.ruleset, { level: "xml" }),
      rule: decode(attrs.rule, { level: "xml" }),
      priority: decode(attrs.priority, { level: "xml" }),
      message: "",
    };
  }

  protected onTagClose(tag: TagCloseNode): void {
    if (tag.name !== "violation" || !this.violation) {
      return;
    }

    this.violation.message = this.violation.message.trim();
    if (!this.violation.message.endsWith(".")) {
      this.violation.message += ".";
    }
    this.report.violations.push(this.violation);
    this.violation = undefined;
  }

  protected onText(tag: TextNode | CDATANode): void {
    this.violation && (this.violation.message += decode(tag.contents, { level: "xml" }));
  }
}

interface FileAttrs {
  name: string;
}

interface ViolationAttrs {
  beginline: string;
  endline: string;
  begincolumn: string;
  endcolumn: string;
  rule: string;
  ruleset: string;
  priority: string;
}
