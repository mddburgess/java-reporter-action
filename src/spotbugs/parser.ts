import { decode } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import SpotbugsReport, { SpotbugsBug } from "./types";

export default class SpotbugsParser extends ReportParser<SpotbugsReport> {
  private category = "";
  private bug: SpotbugsBug = {
    filePath: "",
    startLine: 0,
    endLine: 0,
    category: "",
    priority: 0,
    shortMessage: "",
    longMessage: "",
  };

  constructor(reportPath: string) {
    super(
      {
        categories: new Map<string, string>(),
        bugs: [],
      },
      reportPath
    );
  }

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "BugInstance":
        this.onBugInstanceOpen(parseAttrs(tag.attrs) as BugInstanceAttrs);
        break;
      case "SourceLine":
        this.onSourceLineOpen(parseAttrs(tag.attrs) as SourceLineAttrs);
        break;
      case "BugCategory":
        this.onBugCategoryOpen(parseAttrs(tag.attrs) as BugCategoryAttrs);
        break;
      default:
        break;
    }
  }

  private onBugInstanceOpen(attrs: BugInstanceAttrs) {
    this.bug = {
      filePath: "",
      startLine: 0,
      endLine: 0,
      category: decode(attrs.category, { level: "xml" }),
      priority: Number(attrs.priority),
      shortMessage: "",
      longMessage: "",
    };
    this.report.bugs.push(this.bug);
  }

  private onSourceLineOpen(attrs: SourceLineAttrs) {
    this.bug.filePath = decode(attrs.sourcepath, { level: "xml" });
    this.bug.startLine = Number(attrs.start) || this.bug.startLine;
    this.bug.endLine = Number(attrs.end) || this.bug.startLine;
  }

  private onBugCategoryOpen(attrs: BugCategoryAttrs) {
    this.category = attrs.category;
  }

  protected onTagClose(tag: TagCloseNode): void {
    if (tag.name === "BugCategory") {
      this.category = "";
    }
  }

  protected onText(tag: TextNode | CDATANode): void {
    switch (this.getContext()) {
      case "ShortMessage":
        this.bug.shortMessage = decode(tag.contents, { level: "xml" });
        break;
      case "LongMessage":
        this.bug.longMessage = decode(tag.contents, { level: "xml" });
        break;
      case "Description":
        this.category && this.report.categories.set(this.category, tag.contents);
        break;
      default:
        break;
    }
  }
}

interface BugInstanceAttrs {
  category: string;
  priority: string;
}

interface SourceLineAttrs {
  sourcepath: string;
  start: string;
  end: string;
  startBytecode: string;
  endBytecode: string;
}

interface BugCategoryAttrs {
  category: string;
}
