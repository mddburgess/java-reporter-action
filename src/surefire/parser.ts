import { decode } from "html-entities";
import { parseAttrs } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";
import ReportParser from "../common/parser";
import SurefireReport, { SurefireTestCase, SurefireTestResult } from "./types";

export default class SurefireParser extends ReportParser<SurefireReport> {
  private testCase: SurefireTestCase = {
    className: "",
    testName: "",
    result: "success",
  };

  public constructor(reportPath: string) {
    super(
      {
        name: "",
        tests: 0,
        failures: 0,
        errors: 0,
        skipped: 0,
        testCases: [],
      },
      reportPath
    );
  }

  protected onTagOpen(tag: TagOpenNode): void {
    switch (tag.name) {
      case "testsuite":
        this.onTestSuiteOpen(parseAttrs(tag.attrs) as TestSuiteAttrs);
        break;
      case "testcase":
        this.onTestCaseOpen(parseAttrs(tag.attrs) as TestCaseAttrs);
        break;
      case "failure":
      case "error":
      case "skipped":
        this.onTestResultOpen(tag.name, parseAttrs(tag.attrs) as TestResultAttrs);
        break;
      default:
        break;
    }
  }

  private onTestSuiteOpen(attrs: TestSuiteAttrs) {
    this.report.name = attrs.name;
    this.report.tests = Number(attrs.tests);
    this.report.failures = Number(attrs.failures);
    this.report.errors = Number(attrs.errors);
    this.report.skipped = Number(attrs.skipped);
  }

  private onTestCaseOpen(attrs: TestCaseAttrs) {
    this.testCase.className = decode(attrs.classname, { level: "xml" });
    this.testCase.testName = decode(attrs.name, { level: "xml" });
  }

  private onTestResultOpen(result: SurefireTestResult, attrs: TestResultAttrs) {
    this.testCase.result = result;
    this.testCase.message = decode(attrs.message, { level: "xml" });
  }

  protected onTagClose(tag: TagCloseNode): void {
    if (tag.name !== "testcase" || this.testCase.result === "success") {
      return;
    }

    this.testCase.stackTrace = this.testCase.stackTrace?.trim();
    this.report.testCases.push(this.testCase);
    this.testCase = {
      className: "",
      testName: "",
      result: "success",
    };
  }

  protected onText(tag: TextNode | CDATANode): void {
    const context = this.getContext();
    if (context === "failure" || context === "error" || context === "skipped") {
      this.testCase.stackTrace = (this.testCase.stackTrace || "").concat(
        decode(tag.contents, { level: "xml" })
      );
    }
  }
}

interface TestSuiteAttrs {
  name: string;
  tests: string;
  failures: string;
  errors: string;
  skipped: string;
}

interface TestCaseAttrs {
  classname: string;
  name: string;
}

interface TestResultAttrs {
  message: string;
}
