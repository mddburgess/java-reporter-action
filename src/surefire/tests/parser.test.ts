import * as fs from 'fs';
import {SurefireParser} from '../parser';
import {findFile} from '../../common/files';

describe('Surefire parser', () => {

    it('can parse a simple Surefire report', async () => {
        const path = await findFile('**/TEST-com.example.demo.SimpleTest.xml');
        const xml = await fs.promises.readFile(path);

        const surefireReport = new SurefireParser().parse(xml);
        expect(surefireReport).toMatchSnapshot();
    })
})
