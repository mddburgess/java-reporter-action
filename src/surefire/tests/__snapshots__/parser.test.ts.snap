// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surefire parser can parse a simple Surefire report 1`] = `
Object {
  "errors": 2,
  "failures": 2,
  "skipped": 2,
  "testCases": Array [
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": "void com.example.demo.SimpleTest.disabledTest() is @Disabled",
      "result": "skipped",
      "testName": "disabledTest",
    },
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": "error message",
      "result": "error",
      "stackTrace": "java.lang.RuntimeException: error message
	at com.example.demo.SimpleTest.errorTestWithMessage(SimpleTest.java:35)
",
      "testName": "errorTestWithMessage",
    },
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": undefined,
      "result": "failure",
      "stackTrace": "org.opentest4j.AssertionFailedError: 
	at com.example.demo.SimpleTest.failingTestWithoutMessage(SimpleTest.java:29)
",
      "testName": "failingTestWithoutMessage",
    },
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": "assumption was not met due to: &#10;Expecting:&#10; &lt;1&gt;&#10;to be equal to:&#10; &lt;null&gt;&#10;but was not.",
      "result": "skipped",
      "stackTrace": "org.opentest4j.TestAbortedException: 
assumption was not met due to: 
Expecting:
 <1>
to be equal to:
 <null>
but was not.
	at com.example.demo.SimpleTest.skippedTest(SimpleTest.java:51)
Caused by: org.opentest4j.AssertionFailedError: 

Expecting:
 <1>
to be equal to:
 <null>
but was not.
	at com.example.demo.SimpleTest.skippedTest(SimpleTest.java:51)
",
      "testName": "skippedTest",
    },
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": undefined,
      "result": "error",
      "stackTrace": "java.lang.RuntimeException
	at com.example.demo.SimpleTest.errorTestWithoutMessage(SimpleTest.java:40)
",
      "testName": "errorTestWithoutMessage",
    },
    Object {
      "className": "com.example.demo.SimpleTest",
      "message": "&#10;Expecting:&#10; &lt;4&gt;&#10;to be equal to:&#10; &lt;5&gt;&#10;but was not.",
      "result": "failure",
      "stackTrace": "org.opentest4j.AssertionFailedError: 

Expecting:
 <4>
to be equal to:
 <5>
but was not.
	at com.example.demo.SimpleTest.failingTestWithMessage(SimpleTest.java:24)
",
      "testName": "failingTestWithMessage",
    },
  ],
  "tests": 7,
}
`;
